### PR TITLE
Implement token-based truncation using tiktoken

### LINKED ISSUE #250: We should truncate events by tokens, not characters
If you search `maybe_truncate` through the repo, you will find that inside `maybe_truncate`, we truncate by number of characters in a string.

We should consider using a generic tokenizer, like https://github.com/openai/tiktoken to tokenize the message and try to truncate by tokens instead of characters.

### PR DESCRIPTION
**Pull request by xingyaoww**
## Summary

This PR implements token-based truncation instead of character-based truncation for better LLM context management, addressing issue #250.

### Core Implementation
- **Added tiktoken dependency**: Added `tiktoken>=0.8.0` to SDK pyproject.toml for accurate token counting
- **New function**: `maybe_truncate_by_tokens()` - dedicated token-based truncation function
- **Enhanced existing function**: Updated `maybe_truncate()` with `use_tokens` parameter for backward compatibility
- **Constants**: Added `DEFAULT_TOKEN_LIMIT` constant (12,000 tokens)

### Features
- **Multiple encoding support**: Supports different tokenizers (cl100k_base, p50k_base, etc.)
- **Head/tail preservation**: Maintains existing logic for preserving beginning and end of content
- **Graceful fallback**: Falls back to character-based truncation if encoding is unavailable
- **Edge case handling**: Properly handles invalid encodings, oversized notices, and empty content

### Testing
- **Comprehensive test coverage**: Added 12 new test cases covering various scenarios
- **Backward compatibility**: All existing tests continue to pass
- **Edge cases**: Tests for invalid encodings, different tokenizers, and boundary conditions

## Technical Details

The implementation uses OpenAI's tiktoken library to accurately count tokens based on the specific encoding used by different LLM models. This provides much more accurate context management compared to character-based approximations.

### Key Functions

1. **`maybe_truncate_by_tokens()`**: New dedicated function for token-based truncation
2. **`maybe_truncate()`**: Enhanced with `use_tokens=False` parameter for backward compatibility

### Encoding Support
- `cl100k_base` (GPT-4, GPT-3.5-turbo)
- `p50k_base` (GPT-3, Codex)
- Custom encodings as needed

## Backward Compatibility

✅ All existing functionality remains unchanged by default
✅ Existing tests continue to pass
✅ New functionality is opt-in via the `use_tokens` parameter

## Testing

```bash
# Run truncation-specific tests
uv run pytest tests/sdk/utils/test_truncate.py -v

# Run full SDK test suite
uv run pytest tests/sdk/ -v
```

All tests pass successfully (624 passed).

## Fixes

Closes #250

## Checklist

- [x] Added tiktoken dependency
- [x] Implemented token-based truncation functions
- [x] Added comprehensive tests
- [x] Maintained backward compatibility
- [x] Updated exports in `__init__.py`
- [x] All tests pass
- [x] Pre-commit hooks pass
- [x] Documentation in docstrings

@xingyaoww can click here to [continue refining the PR](https://app.all-hands.dev/conversations/d6146160dccf4b0185db415b9b98e7a0)

### CODE REVIEW

**Comment by xingyaoww** [comment]
@OpenHands review diff against main, now we truncate by tokens, we should remove the old config variables of # of characters to truncate

**Comment by xingyaoww** [comment]
@OpenHands review diff from main and add more tests for using the truncate by token util function

File: openhands/sdk/llm/message.py [hunk 1]
```diff
@@ -33,12 +33,8 @@ class TextContent(mcp.types.TextContent, BaseContent):
     def to_llm_dict(self) -> dict[str, str | dict[str, str]]:
         """Convert to LLM API format."""
         text = self.text
-        if len(text) > DEFAULT_TEXT_CONTENT_LIMIT:
-            logger.warning(
-                f"TextContent text length ({len(text)}) exceeds limit "
-                f"({DEFAULT_TEXT_CONTENT_LIMIT}), truncating"
-            )
-            text = maybe_truncate(text, DEFAULT_TEXT_CONTENT_LIMIT)
```
Comment (xingyaoww) [inline openhands/sdk/llm/message.py:41]: @OpenHands we should still throw a warning when length exceed in `to_llm_dict` similar to the modification before (check diff against main).

Also, in util, we should use encoding "o200k_base" which is used by gpt-5. We should enforce tiktoken version > 0.11.0.

Also, make the "tokenizer" a helper function in util as well - as add a docstring explain that we are using GPT-5 and the result is a best effort approximation to other llm provider

File: openhands/sdk/llm/message.py [hunk 2]
```diff
@@ -33,12 +33,26 @@ class TextContent(mcp.types.TextContent, BaseContent):
...
+        try:
+            tokenizer = get_tokenizer()
+            token_count = len(tokenizer.encode(text))
+            if token_count > DEFAULT_TOKEN_LIMIT:
+                logger.warning(
+                    f"TextContent token count ({token_count}) exceeds limit "
+                    f"({DEFAULT_TOKEN_LIMIT}), truncating"
+                )
+        except Exception:
+            # Fallback to character-based warning if tokenizer fails
```
Comment (xingyaoww) [inline openhands/sdk/llm/message.py]: @OpenHands we shouldn't expect tokenizer to fail, remove the exeception

**Comment by xingyaoww** [comment]
It seems to slow things quite a bit, so maybe no
